src/ColinHDev/CPlot/provider/DataProvider.php

```diff
@@ -1291,4 +1306,255 @@ public function savePlotRate(Plot $plot, PlotRate $plotRate) : Generator {
+                /** @var WorldSettings|false $world */
+                $world = yield $this->awaitWorld($record["level"]);
+                if($world === false)
+                    continue;
+
+                // load plot
+                /** @var Plot|null $plot */
+                $plot = yield $this->awaitPlot($record["level"], (int)$record["x"], (int)$record["z"]);
+                if($plot === null)
+                    continue;
+
+                // claim plot
+                $senderData = new PlotPlayer($playerData, PlotPlayer::STATE_OWNER);
+                $plot->addPlotPlayer($senderData);
+                yield from DataProvider::getInstance()->savePlotPlayer($plot, $senderData);
+            }
+            foreach($mergeRecords as $mergeRecord) {
+                // load world
+                /** @var WorldSettings|false $world */
+                $world = yield $this->awaitWorld($mergeRecord["level"]);
+                if($world === false)
+                    continue;
+
+                // load merge plot 1
+                /** @var Plot|null $plot */
+                $plot = yield $this->awaitPlot($mergeRecord["level"], (int)$mergeRecord["originX"], (int)$mergeRecord["originZ"]);
+                if($plot === null)
+                    continue;
+
+                // load merge plot 2
+                /** @var Plot|null $plotToMerge */
+                $plotToMerge = yield $this->awaitPlot($mergeRecord["level"], (int)$mergeRecord["mergedX"], (int)$mergeRecord["mergedZ"]);
+                if($plotToMerge === null)
+                    continue;
+
+                // complete merge logic
+                yield from DataProvider::getInstance()->awaitPlotDeletion($plotToMerge);
+                foreach($plotToMerge->getMergePlots() as $mergePlot){
+                    $plot->addMergePlot($mergePlot);
+                    yield from $this->addMergePlot($plot, $mergePlot);
+                }
+                foreach($plotToMerge->getPlotPlayers() as $mergePlotPlayer) {
+                    $plot->addPlotPlayer($mergePlotPlayer);
+                    yield from $this->savePlotPlayer($plot, $mergePlotPlayer);
+                }
+                foreach ($plotToMerge->getFlags() as $mergeFlag) {
+                    $flag = $plot->getFlagByID($mergeFlag->getID());
+                    if ($flag === null) {
+                        $flag = $mergeFlag;
+                    } else {
+                        $flag = $flag->merge($mergeFlag->getValue());
+                    }
+                    $plot->addFlag($flag);
+                    yield from DataProvider::getInstance()->savePlotFlag($plot, $flag);
+                }
+                foreach ($plotToMerge->getPlotRates() as $mergePlotRate) {
+                    $plot->addPlotRate($mergePlotRate);
+                    yield from DataProvider::getInstance()->savePlotRate($plot, $mergePlotRate);
+                }
+            }
+            foreach($records as $record) {
```

Comment:
Is there a reason you split the record loop?

Reply:
I figure it's better to do the merge operations before all the flags are set rather than after, though I might being too careful to ensure all the data is synced between plots.

Reply:
also not all players are registered into CPlot yet for helper and denied players, so the second loop-through ensures they are known already.